### PR TITLE
Cache opt

### PR DESCRIPTION
**Pull request by tshemsedinov**

### CODE REVIEW

File: lib/api.common.js [hunk 1]
```diff
@@ -594,18 +594,17 @@ api.common.sortCompareByName = (
   return 0;
 };
 
-api.common.clearCacheStartingWith = (
+api.common.clearCacheStartsWith = (
```
Comment (aqrln) [inline lib/api.common.js]: I liked the previous name more. @belochub, what's your opinion?

Comment (aqrln) [inline lib/api.common.js]: OTOH, the previous name may seem to be somehow connected to time, not name prefixes.

Comment (aqrln) [inline lib/api.common.js]: Maybe it is even more appropriate to indicate somehow that _names_ are starting with something.

File: lib/impress.cache.js [hunk 2]
```diff
@@ -133,66 +132,58 @@ impress.cache.mixin = (application) => {
...
-            key !== filePath &&
-            key.startsWith(filePath) &&
-            !files[key]
-          ) {
-            files[key] = stats;
+        // Directory changed
+        let key, starts, includes;
+        for (key of cache.files.keys()) {
+          starts = key.startsWith(filePath);
+          includes = files.includes(key);
```
Comment (aqrln) [inline lib/impress.cache.js]: Maybe `files` can be a `Set`?

**Comment by aqrln** [comment]
I'll review it more thoroughly later.

File: lib/api.common.js (hunk 1 above)
Comment (aqrln) [inline lib/api.common.js]: Ugh, never mind. For some reason I only reviewed the first commit, not the whole PR.

File: lib/impress.cache.js (hunk 2 above)
Comment (tshemsedinov) [inline lib/impress.cache.js]: Sure

File: lib/api.common.js (hunk 1 above)
Comment (tshemsedinov) [inline lib/api.common.js]: I removed this function at all, but Maps coming from factory `api.common.cache()` have `add`, `del` and `clr` methods

Comment (aqrln) [inline lib/api.common.js]: Yeah, I got it, that's what my last comment is about :)

File: lib/api.common.js [hunk 3]
```diff
@@ -657,39 +643,76 @@ api.common.eventEmitter = () => {
 
 api.common.cache = (
   // Extend Map interface total allocated size: map.allocated
+  options
 ) => {
   const cache = new Map();
+  const calcSize = (options && options.calcSize) || false;
```
Comment (aqrln) [inline lib/api.common.js]: I'd use `0` instead of `false` here so that the data types are consistent.

File: lib/api.common.js [hunk 4]
```diff
@@ -657,39 +643,76 @@ api.common.eventEmitter = () => {
...
+
+};
+
+api.common.cache.withSize = (
+  cache // mixin to this Map
+) => {
+
   cache.allocated = 0;
 
+  const dataSize = data => (data && data.lenght ? data.length : 0);
```
Comment (aqrln) [inline lib/api.common.js]: `length`, not `lenght`

File: lib/api.common.js (hunk 3 above)
Comment (aqrln) [inline lib/api.common.js]: Never mind, I'd thought I was a number before I read the code further. Maybe it's indicating that the variable name isn't that good though.

File: lib/api.common.js (hunk 4 above)
Comment (aqrln) [inline lib/api.common.js]: This typo will cause `dataSize` to be always `0`.

File: lib/api.common.js [hunk 5]
```diff
@@ -657,39 +643,76 @@ api.common.eventEmitter = () => {
...
     if (cache.has(key)) {
       const val = cache.get(key);
-      cache.allocated -= api.common.cacheSize(val);
+      cache.allocated -= dataSize(val);
     }
   };
 
-  cache.clearStartingWith = (
-    startsWith // string to compare with key start
+  cache.clr = (
```
Comment (aqrln) [inline lib/api.common.js]: I believe it is duplicate code.

File: lib/api.common.js [hunk 6]
```diff
@@ -657,39 +643,76 @@ api.common.eventEmitter = () => {
...
+    prefix, // string to compare with key start
+    fn // function(key, val) to be called on each key (optional)
   ) => {
-    let key, val;
-    for ([key, val] of cache) {
-      if (key.startsWith(startsWith)) {
-        cache.allocated -= api.common.cacheSize(val);
+    cache.forEach((val, key) => {
```
Comment (aqrln) [inline lib/api.common.js]: Despite that it has inconsistent loop style :)

File: lib/api.common.js (hunk 5 above)
Comment (tshemsedinov) [inline lib/api.common.js]: Not duplicate, forEach and for..of

File: lib/api.common.js (hunk 6 above)
Comment (tshemsedinov) [inline lib/api.common.js]: What are you talking about?

File: lib/api.common.js (hunk 5 above)
Comment (aqrln) [inline lib/api.common.js]: Ah, I didn't notice it mutates `cache.allocated`. Never mind then :)

File: lib/api.common.js (hunk 6 above)
Comment (aqrln) [inline lib/api.common.js]: I thought the only difference was `for-of` vs `forEach`, but it actually needs `val`, so yeah, everything's okay here.

**Pull request merged**